### ISSUE TITLE
Labels buffer flickering

### ISSUE THREAD

**Issue by HWerneck** [Farama-Foundation/ViZDoom#502]
Hey all!

I just ran into some trouble using the labels buffer.
First, it is giving the same value to both monsters, items and the walls.
Second, it is flickering, randomly changing the values of each element (floor, wall, monster, item). It happens specially when the agent shoots, like almost every time, but even so, not certainly.
My code ran very smooth using the basic.wad environment, and the labels buffer seems ok with it. This is happening in a custom map I created.

The piece of code that treats the labels buffer is adapted from the labels.py example (below). If needed, I can provide a bigger part of the code.
I could figure the player is always labelled 255; the labels buffer show walls = 0 and floor/ceiling = 1 in the example provided.
As I could not (don't know how to) get the precise label of each element in the scene, I painted everything red after painting walls and floor, and then would only paint the player white afterwards.

It worked for the basic.wad file, but not for my custom file.
The screen I am getting for both occasions is [here](https://imgur.com/a/xjxnlhy), and it is run with the same code.

```
labels_buffer = self.game.get_state().labels_buffer
frame = np.stack([labels_buffer] * 3, -1)
frame[labels_buffer == 0] = [0, 0, 0]
frame[labels_buffer == 1] = [64, 64, 64]
frame[labels_buffer > 1] = [255, 0, 0]
frame[labels_buffer == 255] = [255, 255, 255]
return frame
```

I'm running it with Python 3.8.10 in a Ubuntu 20.04.4 LTS virtual machine. ViZDoom version is 1.1.11
Thanks in advance for any help.

**Comment by mwydmuch** [comment]
Hi @HWerneck, you may want to check the labels buffer example: https://github.com/mwydmuch/ViZDoom/blob/master/examples/python/labels.py

>  First, it is giving the same value to both monsters, items and the walls.

So this is happening on your custom map, right? Can you share the map, so I can check what can be the cause of the issue?

> Second, it is flickering, randomly changing the values of each element (floor, wall, monster, item). It happens especially when the agent shoots, like almost every time, but even so, not certainly.

This is expected behavior. The labels buffer is not designed to be a stable image. It can change values for the objects in every frame. To identify objects, one needs to refer to the `state.labels` list, which lists all the objects visible in the frame and their assigned values in the labels buffer (value attribute). This is presented in the example linked above. This way, the labels buffer will never run out of the values for the unique objects.

**Comment by HWerneck** [comment]
> So this is happening on your custom map, right? Can you share the map, so I can check what can be the cause of the issue?

[env2_key.zip](https://github.com/mwydmuch/ViZDoom/files/8469790/env2_key.zip)
There you go, two of the custom maps.

> This is expected behavior.

I get it is not meant to be a stable image, but I didn't have the intention of adding labels to the elements. They are not so "out there" as the identification I actually accomplished, and I need to identify them in a more visible way. If I can get unique ids for each element, and walls, floor/ceiling and player keep steady, my script can paint them the same way, even if they flicker.

Thanks so much for the help, @mwydmuch!!, and sorry my delay.

**Comment by Miffyli** [comment]
> I get it is not meant to be a stable image, but I didn't have the intention of adding labels to the elements. They are not so "out there" as the identification I actually accomplished, and I need to identify them in a more visible way. If I can get unique ids for each element, and walls, floor/ceiling and player keep steady, my script can paint them the same way, even if they flicker.

I am not sure quite what you mean, but let me further explain what Marek told.

The numbers in `labels_buffer` are *not* IDs of objects; the numbers tell which of objects from `labels` (note that this is a different buffer!) is rendered at that pixel. 

[Here](https://github.com/mwydmuch/ViZDoom/blob/master/gym_wrapper/base_gym_env.py#L206-L215) is an another code which renders different objects with different (but stable) colors. Notice the multiple steps of processing:

1) Iterate through objects in the `labels` list (not `labels_buffer`)
2) For each item, choose a color based on `object_id`. _This_ ID is more-or-less same throughout the existence of the object.
3) Now in `labels_buffer` (the image-like array), find where the number corresponds to `label.value`, and assign the item's color to those locations.

In other words, `labels_buffer` are numbers which correspond to `value` attribute of labels in the `labels` buffer, and you need to perform this connection manually.

**Comment by HWerneck** [comment]
Hey @Miffyli! Thanks for the clarification.
I did know they were not IDs of objects, I used the word more like "a number that identifies the object being rendered at that pixel", but didn't think their value actually represented anything.
It takes a little more fiddling, but it is great to know I can consistently paint them on-screen. Nice! Thanks a lot for that.

Anyways, like I said, my script differentiate objects enough for the test (although I think I'm trying your idea), but most objects are being rendered with the same value, hence why the image I sent represents objects with the same color as the floor or the walls.

I will try your idea and report back, even if I am not using for my work.

**Comment by mwydmuch** [comment]
Hi @HWerneck, I'm sorry for the long response time about the problem with labels being colored to 0. I wrote the test for detecting such cases (https://github.com/mwydmuch/ViZDoom/blob/master/tests/test_labels_buffer.py). I run many episodes with standard scenarios, with your provided maps, and also I played ~10 minutes of the standard game myself to see if this kind of error will be detected. I did it on macOS and Ubuntu 20.04 and never encountered the case you presented on your screenshots. So I couldn't replicate the issue so far. Can you provide also the exact script that causes this issue for you? Maybe it's a result of some specific setting? Could you also provide information about your gcc compiler?

**Comment by mwydmuch** [comment]
No further response from the author of the issue, closing.
